flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/WatermarkITCase.java

```diff
@@ -135,6 +135,27 @@ void after() throws Exception {
         }
     }
 
+    private JobID submitJobAndWaitForOperator3Initialized(
+            StreamGraph streamGraph, boolean watermarkIsLongType)
+            throws ExecutionException, InterruptedException {
+        // block operator2's subtask 0 and subtask 1
+        Operator2ProcessFunction.blockSubTasks(0, 1);
+
+        // submit job
+        JobID jobId = flinkCluster.submitJob(streamGraph).get().getJobID();
+
+        // wait all operator3 tasks have initialized completely
+        tryWaitUntilCondition(
+                () -> Operator3ProcessFunction.receivedWatermarks.size() == DEFAULT_PARALLELISM);
```

Comment:
Here is the main fix, by waiting for `receivedWatermarks` size, not `attemptIds` size.